xds/src/main/java/io/grpc/xds/internal/certprovider/MeshCaCertificateProvider.java

```diff
@@ -17,35 +17,349 @@
+    }
+    getWatcher().onError(Status.fromThrowable(t));
+  }
+
+  private KeyPair generateKeyPair() throws NoSuchAlgorithmException {
+    KeyPairGenerator keyPairGenerator = KeyPairGenerator.getInstance(alg);
+    keyPairGenerator.initialize(keySize);
+    return keyPairGenerator.generateKeyPair();
+  }
+
+  private String generateCsr(KeyPair pair) throws IOException, OperatorCreationException {
+    PKCS10CertificationRequestBuilder p10Builder =
+        new JcaPKCS10CertificationRequestBuilder(
+            new X500Principal("CN=EXAMPLE.COM"), pair.getPublic());
+    JcaContentSignerBuilder csBuilder = new JcaContentSignerBuilder(signatureAlg);
+    ContentSigner signer = csBuilder.build(pair.getPrivate());
+    PKCS10CertificationRequest csr = p10Builder.build(signer);
+    PemObject pemObject = new PemObject("NEW CERTIFICATE REQUEST", csr.getEncoded());
+    StringWriter str = new StringWriter();
+    JcaPEMWriter pemWriter = new JcaPEMWriter(str);
+    pemWriter.writeObject(pemObject);
+    pemWriter.close();
+    str.close();
+    return str.toString();
+  }
+
+  /** Compute refresh interval as half of interval to current cert expiry. */
+  private long computeRefreshSecondsFromCurrentCertExpiry() {
+    X509Certificate lastCert = getWatcher().getLastIdentityCert();
+    if (lastCert == null) {
+      return INITIAL_DELAY_SECONDS;
+    }
+    long delayToCertExpirySeconds = computeDelaySecondsToCertExpiry(lastCert) / 2;
+    return Math.max(delayToCertExpirySeconds, INITIAL_DELAY_SECONDS);
+  }
+
+  private long computeDelaySecondsToCertExpiry(X509Certificate lastCert) {
+    checkNotNull(lastCert, "lastCert");
+    return TimeUnit.NANOSECONDS.toSeconds(
+        TimeUnit.MILLISECONDS.toNanos(lastCert.getNotAfter().getTime()) - timeProvider
+            .currentTimeNanos());
+  }
+
+  private static void shutdownChannel(ManagedChannel channel) {
+    channel.shutdown();
+    try {
+      channel.awaitTermination(10, TimeUnit.SECONDS);
+    } catch (InterruptedException ex) {
+      logger.log(Level.SEVERE, "awaiting channel Termination", ex);
+      channel.shutdownNow();
+    }
+  }
+
+  private List<X509Certificate> getX509CertificatesFromResponse(
+      Meshca.MeshCertificateResponse response) throws CertificateException, IOException {
+    List<String> certChain = response.getCertChainList();
+    List<X509Certificate> x509Chain = new ArrayList<>(certChain.size());
+    for (String certString : certChain) {
+      x509Chain.add(
+          CertificateUtils
+              .toX509Certificate(new ByteArrayInputStream(certString.getBytes(UTF_8))));
+    }
+    return x509Chain;
+  }
+
+  @VisibleForTesting
+  class RefreshCertificateTask implements Runnable {
+    @VisibleForTesting final SynchronizationContext.ScheduledHandle scheduledHandle;
+
+    private RefreshCertificateTask(long delayInSeconds) {
+      scheduledHandle =
+          syncContext.schedule(this, delayInSeconds, TimeUnit.SECONDS, scheduledExecutorService);
```

Comment:
It might be better to move it to the `scheduleNextRefreshCertificate()` method.:

~`@VisibleForTesting RefreshCertificateTask scheduledTask;`~
```java
@VisibleForTesting SynchronizationContext.ScheduledHandle scheduledHandle; // top level field

private void scheduleNextRefreshCertificate(long delayInSeconds) {
  if (scheduledHandle != null && scheduledHandle.isPending()) {
    logger.log(Level.SEVERE, "Pending task found: inconsistent state in scheduledHandle!");
    scheduledHandle.cancel();
  }
  RefreshCertificateTask runnable = new RefreshCertificateTask();  // Pure Runnable class
  scheduledHandle = syncContext.schedule(
      runnable, delayInSeconds, TimeUnit.SECONDS, scheduledExecutorService);
}

class RefreshCertificateTask implements Runnable {
  @Override
  public void run() {
    try {
      refreshCertificate();
    } catch (NoSuchAlgorithmException | OperatorCreationException | IOException ex) {
      logger.log(Level.SEVERE, "refreshing certificate", ex);
    }
  }
}
```

Reply:
That's what I had earlier and I understand that's the pattern followed elsewhere. There was a reason I made this change but if that reason doesn't exist I'll implement what you suggested.

Reply:
Done